clar2wasm/src/words/sequences.rs

```diff
@@ -528,6 +528,44 @@ impl ComplexWord for Map {
     ) -> Result<(), GeneratorError> {
         let fname = args.get_name(0)?;
 
+        let seq_ty = generator
+            .get_expr_type(args.get_expr(1)?)
+            .ok_or_else(|| GeneratorError::TypeError("list expression must be typed".to_owned()))?
+            .clone();
+
+        // WORKAROUND: Get the type of the function being called, and set the
+        // type of the sequence value to match the functions parameter type.
+        // This is a workaround for the typechecker not being able to infer
+        // the complete type of initial value.
+        if let TypeSignature::SequenceType(SequenceSubtype::ListType(lt)) = &seq_ty {
```

Comment:
nit: from a quick skim this seems similar in spirit to what `typechecker_workaround` does. If so, should we move this workaround in there as well?

Reply:
Good call @matteojug ! It happens to be the same workaround used for `fold`. 

But we still need to revisit the code to work on the remaining workarounds. For instance the `fold` is still using a [local workaround](https://github.com/stacks-network/clarity-wasm/blob/main/clar2wasm/src/words/sequences.rs#L106-L116) plus one in the `typechecker_workaround` function.

Opened #491 to track that.

Reply:
We may need to fix `typechecker_workaround` as currently it only operates on top level expressions, thus ignoring nested ones:
```rust
#[test]
fn crosscheck_map_ok_buff(buf in buffer(50)) {
    let snippet = format!(r#"
    (define-private (foo (a (response (buff 50) int))) (len (unwrap! a u0)))
    (begin (map foo (list (ok {buf})))) ;; added a wrapping `begin` here to an otherwise passing test
    "#);

    crosscheck(
        &snippet,
        Ok(Some(Value::cons_list_unsanitized(vec![Value::UInt(50)]).unwrap()))
    )
}
```
```
Err(Wasm(UnableToLoadModule(WebAssembly translation error

Caused by:
    Invalid input WebAssembly code at offset 7800: type mismatch: expected i64, found i32
```

This may be impacting also `filter`/`fold`.

If the previous version with the local logic was working we could first revert to that to merge this, and fix `typechecker_workaround` (moving back there also the map logic) in a separate PR.

Reply:
Local workaround is working also for nested. Reverted.

Reply:
@matteojug It's not the same as `typechecker_workaround`. It's a bit subtle, but:

- These local workarounds are for when you have the correct type for the expression, but one of the arguments is incorrectly typed. In this case, the final type of `map` was correctly computed.
- `typechecker_workaround` is for when the entire expression has a wrong type, making it impossible to deduce the correct type of the arguments. In this case we have no choice but to re-parse the source code to set the correct type for the expression.

Reply:
Got the gist of it, thanks!